Evolutionary_loop/Loop_Parts/Part_B/part_b_job1.py

```diff
@@ -0,0 +1,229 @@
+        516.68, 533.34, 550.01, 566.68, 583.34, 600.01, 616.68, 633.34, 650.01,
+        666.68, 683.35, 700.01, 716.68, 733.35, 750.01, 766.68, 783.35, 800.01,
+        816.68, 833.35, 850.02, 866.68, 883.35, 900.02, 916.68, 933.35, 950.02,
+        966.68, 983.35, 1000.00, 1016.70, 1033.40, 1050.00, 1066.70] 
+
+    sim_path = xmacros_dir/ 'simulation_PEC.xmacro'
+    # If simulation.xmacro exists, empty it (without changing permission)
+    sp.run(f'[ -f {sim_path} ] && > {sim_path}', shell=True)
+    
+    # Begin writing some run-specific variables to the xmacro
+    with open (sim_path, "w+") as f:
+        f.write(f'var NPOP = {npop};\n'
+                f'var indiv = {indiv};\n'
+                f'//Factor of {geo_factor} frequency\n'
+                'var freq = [')
+        f.write(', '.join( "%.2f"%(freq*geo_factor) for freq in freq_list) )
+        f.write('] \n')
+    
+    if (gen == 0 and indiv == 1):
+        with open (sim_path,"a") as f:
+            f.write('if(indiv==1){\n'
+                    f'App.saveCurrentProjectAs(\"{working_dir}/Run_Outputs/'
+                    f'{run_name}/{run_name}\");\n'
+                    '}\n')
+
+    os.umask(0)                # umask(0) needed to ensure chmod works properly.
+    os.chmod(sim_path, 0o775 ) # This is the same as Bash chmod 775 ${sim_path}.
+
+
+## CAT SECTION
+
+    # The rest of simulation_PEC.xmacro is built from two skeleton text files.
+    if curved == 0:                            
+        if nsections == 1:                     # straight-side symmetric bicone
+            sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton_GPU.txt'
+                   f'>> {sim_path}', shell=True)
+            sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton2_GPU.txt'
+                   f'>> {sim_path}', shell=True)
+
+        else:                                  # straight-side asymmetric bicone
+            sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton_Sep.txt'
+                   f'>> {sim_path}', shell=True)
+            sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton2_Sep.txt'
+                   f'>> {sim_path}', shell=True)
+    else:                                      # curved-side bicone
+        sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton_curved.txt'
+               f'>> {sim_path}', shell=True)
+        sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton2_curved.txt'
+               f'>> {sim_path}', shell=True)
+
+
+## SED-TION (Ha! Get it? Like...section?)
+
+    # sed swaps out the hardcoded word "fileDirectory" inside the xmacro.
+    # option "i" of sed: swap "in-place" instead of outputting to terminal.
+    # The empty quotes after sed -i are for MacOS compatibility:
+    # https://stackoverflow.com/questions/12272065/sed-undefined-label-on-macos
+    sp.run(f'sed -i "" "s|fileDirectory|{working_dir}/Generation_Data|" '
+           f'{sim_path}', shell=True)
```

Comment:
If you set the var at the top like we discussed, you can get rid of this.

Otherwise you can replace `fileDirectory` with `{fileDirectory}` in the file, and then do something like:
```py
content = open(path).read_text()

with open(path) as f:
    f.write(content.format(fileDirectory=str(working_dir)))
```
This will open the file, read the content, replace the value using `.format()` and rewrite it.

Reply:
I think it is perhaps to leave the macroskeleton `.txt` files alone for now, as I heard that @Machtay is working on editing those.
I tried the second approach since it involves fewer edits on these macroskeletons (ie swapping out `filedirectory` with `{fileDirectory}`, and then
`  content = sim_path.read_text()
    with open(sim_path, 'w') as f:
        f.write(content.format(fileDirectory='HELLO WORLD I AM AN ERROR'))`

which resulted in the following key error:
`Traceback (most recent call last):
  File "/Users/Jason/Documents/OSU/GENETIS/GENETIS_HPol/Evolutionary_loop/Loop_Parts/Part_B/part_b_job1.py", line 219, in <module>
    main(args.indiv, args.gen, args.npop, args.working_dir, args.run_name,
  File "/Users/Jason/Documents/OSU/GENETIS/GENETIS_HPol/Evolutionary_loop/Loop_Parts/Part_B/part_b_job1.py", line 141, in main
    f.write(content.format(fileDirectory='HELLO WORLD I AM AN ERROR'))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: '\nApp'
`